File: python/lib/processor.py
import re

def top_resources(logs, top=10):
    results = {}
    for line in logs.split('\n'):
        if len(line.strip()) > 0:
            m = re.search(r"\[(.*)\] \"(.*)\" (\d{3}) (\d*)", line)
            [method, resource, protocol] = m.group(2).split(' ')
            status, bytes = int(m.group(3)), int(m.group(4))
            if method == 'GET' and status == 200:
                if resource not in results:
                    results[resource] = (0, 0)
                results[resource] = (results[resource][0] + 1, #occurrences
                                     results[resource][1] + bytes) #bytes
    return sorted(results.items(), key=lambda x: (-x[1][0], -x[1][1], x[0]))[0:top]

File: javascript/lib/processor.js
const re = /\[(.*)\]\s"(.*)"\s(\d{3})\s(\d*)/

const topResources = (logs, top=10) => {
  const results = {}
  for(let line of logs.split('\n')) {
    if(line.trim().length) {
      const [,, req, status, bytes,,] = re.exec(line.trim())
      const [method, resource,] = req.split(' ')
      if(method === 'GET' && status === '200') {
        if(!(resource in results)) {
          results[resource] = {bytes: 0, occurrences: 0}
        }
        results[resource] = {
          bytes: results[resource].bytes + parseInt(bytes),
          occurrences: results[resource].occurrences + 1
        }
      }
    }
  }
  const sortedLogs = Object.keys(results).sort( (a, b) => {
    if(results[b].occurrences > results[a].occurrences) return 1
    if(results[b].occurrences < results[a].occurrences) return -1
    if(results[b].bytes > results[a].bytes) return 1
    if(results[b].bytes < results[a].bytes) return -1
    if(b > a) return -1
    return 1
  })
  return sortedLogs.map(k => {return {[k]: results[k]}}).slice(0, top)
}

module.exports = topResources
